### ISSUE TITLE
SWT TableEditor : Caret is not shown for StyledText Control

### ISSUE THREAD

**Issue by raghucssit** [eclipse-platform/eclipse.platform.swt#644]
**Description**
If we use StyledText control in SWT TableEditor then caret is not visible inside Text Control. But for normal Text control it is visible.

**How to Reproduce**
Create a simple Table editor and use StyledText control for TableEditor.
Please use the attached snippet to reproduce.
[Snippet124withStyledText.txt](https://github.com/eclipse-platform/eclipse.platform.swt/files/11331049/Snippet124withStyledText.txt)

**Expected behavior**
Caret should be visible when user clicks on StyledText Control

**Screenshots**
![caret-missing](https://user-images.githubusercontent.com/10301131/234529177-6d4de53c-4600-4630-8b3e-8fad2fb64c50.png)

**Environment:**
- - [ ] All OS
- - [ ] Windows
- - [x] Linux
- - [x] macOS 
 
**Version**
 swt-3.121

**Comment by Phillipus** [comment]
I tested this on Mac and I can see the caret when I click on the styled text control. However, I have to click twice to see the caret on the simple text controls.

**Comment by raghucssit** [comment]
[TableEditorStyledTextExample.txt](https://github.com/eclipse-platform/eclipse.platform.swt/files/11381605/TableEditorStyledTextExample.txt)
I tried to add `Canvas` control as a table editor. Even on canvas `Caret` does not appear.
This hints that the issue is with `Table` instead of the `Control` which holds the `Caret`. Because i have added same control on a composite  and `Caret` appears.
I am trying to find root cause but so far no success,
@SyntevoAlex Any suggestion/hints ?

**Comment by deepika-u** [comment]
> I tested this on Mac and I can see the caret when I click on the styled text control.

On windows 11 as well, we dont see any problem.

**Comment by deepika-u** [comment]
@akurtakov 
Please have a look. Thanks

**Comment by raghucssit** [comment]
> 
I have created a another simplest example uses only one cell and has canvas.
[SimplestExampleWithOnlyCanvas.txt](https://github.com/eclipse-platform/eclipse.platform.swt/files/11405355/SimplestExampleWithOnlyCanvas.txt)
I tried to find out when it was last worked. It was working on `Version: 2019-03 (4.11)
Build id: I20190307-0500`. In the next build (4.12) editor seems to be completely broken, i.e, editor itself did not render.
I suspect issue is with `Table#setParentGdkResource (Control child)` as it is trying to paint children. But i could not figure out the exact issue.

**Comment by Phillipus** [comment]
I can confirm the caret not showing on Linux.

> I suspect issue is with `Table#setParentGdkResource (Control child)` as it is trying to paint children. But i could not figure out the exact issue.

The code in that method used to be:

```java
if (GTK.GTK_VERSION >= OS.VERSION(3, 10, 0)) {
	hasChildren = true;
	connectFixedHandleDraw();
}
```

And since https://github.com/eclipse-platform/eclipse.platform.swt/commit/c1c3699c9190461b6b64f6c2672c5befd67c3f0c is now:

```java
hasChildren = true;
connectFixedHandleDraw();
```
	
If those two lines are commented out the caret appears.

**Comment by akurtakov** [comment]
Do you say you're running on GTK version older than 3.10.0 ? I don't see how https://github.com/eclipse-platform/eclipse.platform.swt/commit/c1c3699c9190461b6b64f6c2672c5befd67c3f0c makes a difference as SWT requires far newer GTK version.

**Comment by Phillipus** [comment]
> Do you say you're running on GTK version older than 3.10.0

I'm not.

Perhaps back in 2019 users were on an older GTK version so those two lines did not execute. Now that they do execute there is no caret. See 4b0412b94e47be8f2c5480568a36ec71809a148f

**Comment by iloveeclipse** [comment]
> Do you say you're running on GTK version older than 3.10.0 ? I don't see how [c1c3699](https://github.com/eclipse-platform/eclipse.platform.swt/commit/c1c3699c9190461b6b64f6c2672c5befd67c3f0c) makes a difference as SWT requires far newer GTK version.

No, just commenting out these lines on master fixes caret issue in table editors.
So now one has to understand what these lines actually do :-)

**Comment by akurtakov** [comment]
Whoever works on this one - please test every change with full Eclipse as commenting these lines most likely break something else.

**Comment by iloveeclipse** [comment]
> So now one has to understand what these lines actually do :-)

Funny enough, they were added via https://bugs.eclipse.org/bugs/show_bug.cgi?id=531928 via the bug created by us (Advantest) to fix the table editing...

**Comment by raghucssit** [comment]
> unstable

commenting those 2 lines breaks tthe able editor. It renders only first cell, though i added 2 rows and 3 columns.

**Comment by Phillipus** [comment]
> commenting those 2 lines breaks tthe table editor. It renders only first cell, though i added 2 rows and 3 columns.

I can get the first Snippet to show the caret in all cells and work OK by commenting out either one or both of these lines:

```java
hasChildren = true;
connectFixedHandleDraw();
```

I'm on latest Linux Mint Cinnamon.

**Comment by iloveeclipse** [comment]
I believe I have a fix, see https://github.com/eclipse-platform/eclipse.platform.swt/pull/663.

@SyntevoAlex, @akurtakov, @ericwill  : would be great if you could look at it.

It just paints the caret directly on cairo instead of requesting `gtk_widget_queue_draw` to happen for the parent (which seem to do nothing if the parent is embedded in the table).
